OIDC server is erroneously shown as not available

### Describe the bug

I have an app, which uses `quarkus-oidc` and connects to keycloak server. When I deploy the app and keycloak server to OpenShift instance, app fails with "OIDC Server is not available" despite the server being available from the app pod

### Expected behavior

App establishes a connection to keycloak server

### Actual behavior

Acessing the app returns error page, pod logs contain this:
```
2024-02-15 16:09:54,179 WARN  [io.qua.oid.com.run.OidcCommonUtils] (vert.x-eventloop-thread-0) Discovery has failed, status code: 403
2024-02-15 16:09:54,180 WARN  [io.qua.oid.com.run.OidcCommonUtils] (vert.x-eventloop-thread-0) OIDC Server is not available:: io.quarkus.oidc.common.runtime.OidcEndpointAccessException
	at io.quarkus.oidc.common.runtime.OidcCommonUtils.lambda$discoverMetadata$2(OidcCommonUtils.java:450)
	at io.smallrye.context.impl.wrappers.SlowContextualFunction.apply(SlowContextualFunction.java:21)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform$UniOnItemTransformProcessor.onItem(UniOnItemTransform.java:36)
	at io.smallrye.mutiny.vertx.AsyncResultUni.lambda$subscribe$1(AsyncResultUni.java:35)
	at io.smallrye.mutiny.vertx.DelegatingHandler.handle(DelegatingHandler.java:25)
	at io.vertx.ext.web.client.impl.HttpContext.handleDispatchResponse(HttpContext.java:397)
	at io.vertx.ext.web.client.impl.HttpContext.execute(HttpContext.java:384)
	at io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:362)
	at io.vertx.ext.web.client.impl.HttpContext.fire(HttpContext.java:329)
	at io.vertx.ext.web.client.impl.HttpContext.dispatchResponse(HttpContext.java:291)
	at io.vertx.ext.web.client.impl.HttpContext.lambda$null$7(HttpContext.java:507)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:276)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:258)
	at io.vertx.core.impl.ContextInternal.lambda$runOnContext$0(ContextInternal.java:56)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)

2024-02-15 16:09:54,180 WARN  [io.qua.oid.run.OidcRecorder] (vert.x-eventloop-thread-0) OIDC server is not available at the 'http://${URL_EDITED}/realms/test-mutual-tls-realm' URL. Please make sure it is correct. Note it has to end with a realm value if you work with Keycloak, for example: 'https://localhost:8180/auth/realms/quarkus'
2024-02-15 16:09:54,187 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-0) HTTP Request to / failed, error id: 46860153-0184-4e51-bb27-2ce0670cded5-1: io.quarkus.oidc.OIDCException: OIDC Server is not available
	at io.quarkus.oidc.runtime.OidcRecorder.toOidcException(OidcRecorder.java:371)
	at io.quarkus.oidc.runtime.OidcRecorder$11.apply(OidcRecorder.java:502)
	at io.quarkus.oidc.runtime.OidcRecorder$11.apply(OidcRecorder.java:495)
	at io.smallrye.context.impl.wrappers.SlowContextualBiFunction.apply(SlowContextualBiFunction.java:21)
	at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureFlatMap$UniOnItemOrFailureFlatMapProcessor.performInnerSubscription(UniOnItemOrFailureFlatMap.java:86)
	at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureFlatMap$UniOnItemOrFailureFlatMapProcessor.onFailure(UniOnItemOrFailureFlatMap.java:65)
	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onFailure(UniOperatorProcessor.java:55)
	at io.smallrye.mutiny.operators.uni.UniOnFailureTransform$UniOnFailureTransformProcessor.onFailure(UniOnFailureTransform.java:64)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromPublisher$PublisherSubscriber.onError(UniCreateFromPublisher.java:78)
	at io.smallrye.mutiny.helpers.HalfSerializer.onError(HalfSerializer.java:59)
	at io.smallrye.mutiny.helpers.StrictMultiSubscriber.onFailure(StrictMultiSubscriber.java:90)
	at io.smallrye.mutiny.subscription.MultiSubscriber.onError(MultiSubscriber.java:73)
	at io.smallrye.mutiny.subscription.SerializedSubscriber.onFailure(SerializedSubscriber.java:101)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp$RetryWhenOperator.testOnFailurePredicate(MultiRetryWhenOp.java:134)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp$RetryWhenOperator.onFailure(MultiRetryWhenOp.java:117)
	at io.smallrye.mutiny.subscription.MultiSubscriber.onError(MultiSubscriber.java:73)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.onFailure(UniToMultiPublisher.java:102)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform$UniOnItemTransformProcessor.onItem(UniOnItemTransform.java:40)
	at io.smallrye.mutiny.vertx.AsyncResultUni.lambda$subscribe$1(AsyncResultUni.java:35)
	at io.smallrye.mutiny.vertx.DelegatingHandler.handle(DelegatingHandler.java:25)
	at io.vertx.ext.web.client.impl.HttpContext.handleDispatchResponse(HttpContext.java:397)
	at io.vertx.ext.web.client.impl.HttpContext.execute(HttpContext.java:384)
	at io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:362)
	at io.vertx.ext.web.client.impl.HttpContext.fire(HttpContext.java:329)
	at io.vertx.ext.web.client.impl.HttpContext.dispatchResponse(HttpContext.java:291)
	at io.vertx.ext.web.client.impl.HttpContext.lambda$null$7(HttpContext.java:507)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:276)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:258)
	at io.vertx.core.impl.ContextInternal.lambda$runOnContext$0(ContextInternal.java:56)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.RuntimeException: OIDC Server is not available
	at io.quarkus.oidc.common.runtime.OidcCommonUtils.lambda$discoverMetadata$3(OidcCommonUtils.java:459)
	at io.smallrye.context.impl.wrappers.SlowContextualFunction.apply(SlowContextualFunction.java:21)
	at io.smallrye.mutiny.operators.uni.UniOnFailureTransform$UniOnFailureTransformProcessor.onFailure(UniOnFailureTransform.java:54)
	... 29 more
````

### How to Reproduce?

1.  `git@github.com:fedinskiy/reproducer.git -b rhsso-no-server`
2. Deploy keycloak:
```
oc create project reproducer
oc create configmap keycloak-realm --from-file keycloak-realm.json
oc apply -f keycloak.yaml
oc expose svc/rhsso --port=8080
oc scale deployment/rhsso --replicas=1
```
3. Deploy the app:
```
route=$(oc get route rhsso -o template --template='{{.spec.host}}')
mvn clean package -Dquarkus.oidc.auth-server-url=http://${route}/realms/test-mutual-tls-realm
```
4. Check deployment:
```
$ oc get pods,routes
NAME                         READY   STATUS      RESTARTS   AGE
pod/app-1-build              0/1     Completed   0          35m
pod/app-776776dcd4-2p58r     1/1     Running     0          16m
pod/rhsso-759d5ddc55-82hn9   1/1     Running     0          76m

NAME                             HOST/PORT                                          PATH   SERVICES   PORT   TERMINATION   WILDCARD
route.route.openshift.io/app     app-reproducer.apps.${EDITED}            app        http                 None
route.route.openshift.io/rhsso   rhsso-reproducer.apps.${EDITED}          rhsso      8080                 None
```
5. Try to access app:
`$ curl http://app-reproducer.apps.${EDITED} 
{"details":"Error id 46860153-0184-4e51-bb27-2ce0670cded5-2","stack":""}`
6. Check app logs: `$ oc logs app-776776dcd4-2p58r` for the error above
7. Check, that keycloak server is accessible:
```
$ curl http://${route}/realms/test-mutual-tls-realm
{"realm":"test-mutual-tls-realm","public_key":"$EDITED","token-service":"http://rhsso-reproducer.${EDITED}/realms/test-mutual-tls-realm/protocol/openid-connect","account-service":"http://rhsso-reproducer.${EDITED}realms/test-mutual-tls-realm/account","tokens-not-before":0}
```
8. Check, that  keycloak server is accessible from the app pod:
```
echo http://${route}/realms/test-mutual-tls-realm # copy
$ oc rsh pod/app-776776dcd4-2p58r
$ curl $PASTED
{"realm":"test-mutual-tls-realm","public_key":"$EDITED","token-service":"http://rhsso-reproducer.${EDITED}/realms/test-mutual-tls-realm/protocol/openid-connect","account-service":"http://rhsso-reproducer.${EDITED}/realms/test-mutual-tls-realm/account","tokens-not-before":0}
```

### Output of `uname -a` or `ver`

6.7.3-200.fc39.x86_64

### Output of `java -version`

Java version: 21.0.1, vendor: Eclipse Adoptium

### Quarkus version or git rev

 f5d3e6e6b0a9ae4b6c08f789bd655012107f2296

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

 Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)

### Additional information

```
$ oc version
Client Version: 4.14.0-202311021650.p0.g9b1e0d2.assembly.stream-9b1e0d2
Kustomize Version: v5.0.1
Kubernetes Version: v1.27.10+28ed2d7
```

--- a/extensions/oidc-common/runtime/src/main/java/io/quarkus/oidc/common/runtime/OidcCommonUtils.java
+++ b/extensions/oidc-common/runtime/src/main/java/io/quarkus/oidc/common/runtime/OidcCommonUtils.java
@@ -446,7 +446,13 @@ public class OidcCommonUtils {
             if (resp.statusCode() == 200) {
                 return resp.bodyAsJsonObject();
             } else {
-                LOG.warnf("Discovery has failed, status code: %d", resp.statusCode());
+                String errorMessage = resp.bodyAsString();
+                if (errorMessage != null && !errorMessage.isEmpty()) {
+                    LOG.warnf("Discovery request %s has failed, status code: %d, error message: %s", discoveryUrl,
+                            resp.statusCode(), errorMessage);
+                } else {
+                    LOG.warnf("Discovery request %s has failed, status code: %d", discoveryUrl, resp.statusCode());
+                }
                 throw new OidcEndpointAccessException(resp.statusCode());
             }
         }).onFailure(oidcEndpointNotAvailable())

--- a/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcProviderClient.java
+++ b/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcProviderClient.java
@@ -22,7 +22,6 @@ import io.quarkus.oidc.common.OidcRequestFilter;
 import io.quarkus.oidc.common.runtime.OidcCommonConfig.Credentials.Secret.Method;
 import io.quarkus.oidc.common.runtime.OidcCommonUtils;
 import io.quarkus.oidc.common.runtime.OidcConstants;
-import io.quarkus.oidc.common.runtime.OidcEndpointAccessException;
 import io.smallrye.mutiny.Uni;
 import io.smallrye.mutiny.groups.UniOnItem;
 import io.vertx.core.Vertx;
@@ -109,7 +108,7 @@ public class OidcProviderClient implements Closeable {
         if (resp.statusCode() == 200) {
             return new JsonWebKeySet(resp.bodyAsString(StandardCharsets.UTF_8.name()));
         } else {
-            throw new OidcEndpointAccessException(resp.statusCode());
+            throw responseException(metadata.getJsonWebKeySetUri(), resp);
         }
     }
 
@@ -201,7 +200,7 @@ public class OidcProviderClient implements Closeable {
     }
 
     private AuthorizationCodeTokens getAuthorizationCodeTokens(HttpResponse<Buffer> resp) {
-        JsonObject json = getJsonObject(resp);
+        JsonObject json = getJsonObject(metadata.getAuthorizationUri(), resp);
         final String idToken = json.getString(OidcConstants.ID_TOKEN_VALUE);
         final String accessToken = json.getString(OidcConstants.ACCESS_TOKEN_VALUE);
         final String refreshToken = json.getString(OidcConstants.REFRESH_TOKEN_VALUE);
@@ -209,35 +208,41 @@ public class OidcProviderClient implements Closeable {
     }
 
     private UserInfo getUserInfo(HttpResponse<Buffer> resp) {
-        return new UserInfo(getString(resp));
+        return new UserInfo(getString(metadata.getUserInfoUri(), resp));
     }
 
     private TokenIntrospection getTokenIntrospection(HttpResponse<Buffer> resp) {
-        return new TokenIntrospection(getString(resp));
+        return new TokenIntrospection(getString(metadata.getIntrospectionUri(), resp));
     }
 
-    private static JsonObject getJsonObject(HttpResponse<Buffer> resp) {
+    private static JsonObject getJsonObject(String requestUri, HttpResponse<Buffer> resp) {
         if (resp.statusCode() == 200) {
             LOG.debugf("Request succeeded: %s", resp.bodyAsJsonObject());
             return resp.bodyAsJsonObject();
         } else {
-            throw responseException(resp);
+            throw responseException(requestUri, resp);
         }
     }
 
-    private static String getString(HttpResponse<Buffer> resp) {
+    private static String getString(String requestUri, HttpResponse<Buffer> resp) {
         if (resp.statusCode() == 200) {
             LOG.debugf("Request succeeded: %s", resp.bodyAsString());
             return resp.bodyAsString();
         } else {
-            throw responseException(resp);
+            throw responseException(requestUri, resp);
         }
     }
 
-    private static OIDCException responseException(HttpResponse<Buffer> resp) {
+    private static OIDCException responseException(String requestUri, HttpResponse<Buffer> resp) {
         String errorMessage = resp.bodyAsString();
-        LOG.debugf("Request has failed: status: %d, error message: %s", resp.statusCode(), errorMessage);
-        throw new OIDCException(errorMessage);
+
+        if (errorMessage != null && !errorMessage.isEmpty()) {
+            LOG.errorf("Request %s has failed: status: %d, error message: %s", requestUri, resp.statusCode(), errorMessage);
+            throw new OIDCException(errorMessage);
+        } else {
+            LOG.errorf("Request %s has failed: status: %d", requestUri, resp.statusCode());
+            throw new OIDCException("Error status:" + resp.statusCode());
+        }
     }
 
     @Override

